[Merged by Bors] - feat: Interactive Library Rewrite

JovanGerb opened the pull request:
This PR defines the `rw??` library rewrite tactic that uses my `RefinedDiscrTree`.

This is a point&click tactic, letting the user click on subexpressions in the main goal. The results are sorted by relevance, and are separated into sections based on the matched pattern. It comes with a filter button that allows you to switch between the (default) filtered view, with removed duplicate lemmas and not showing lemma names, and the complete view. There is also a section for definitionally unfolding the selected expression - these rewrites do not correspond to a library lemma.

---


- [x] depends on: #11968
- [x] depends on: #12016
- [x] depends on: #25955


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

fpvandoorn commented:
please mark as `awaiting-review` if it's ready for review

JovanGerb commented:
awaiting-review

JovanGerb commented:
!bench

The lazy version is not persistent, so if this bench shows that, I'll make a `StrictRefinedDiscrTree`.

JovanGerb commented:
!bench

JovanGerb commented:
!bench

eric-wieser commented on Mathlib/Tactic/Widget/LibraryRewrite.lean:
```diff
@@ -0,0 +1,645 @@
+def isMVarSwap (t s : Expr) : Bool :=
+  go t s {} |>.isSome
+where
+  /-- The main loop of `isMVarSwap`. Returning `none` corresponds to a failure. -/
```
Can you explain what `some l` means here?

eric-wieser commented on Mathlib/Tactic/Widget/LibraryRewrite.lean:
```diff
@@ -0,0 +1,645 @@
+
+  /-- Render the list of rewrite results in one section. -/
+  renderSectionCore (showNames : Bool) (sec : Array RewriteInterface) : Html :=
+    .element "ul" #[("style", json% { "padding-left" : "30px"})] <|
```
I think you can use `<ul style={}>{... sec.map _}</ul>` or similar here.

JovanGerb commented:
The notable changes I made most recently:
- printing metavariables: The metavariables created by the rewrite are printed as `?_` in the produced `rw` tactic. But in the interface, they are printed as normal (so it is `?a` if the variable is called `a` in the rewrite lemma)
- In a separate file, I added an environment extension for blacklisted imports. I tagged `Init.Omega`, `Init.Grind` and `Mathlib.Tactic`. So, `rw??` excludes lemmas from there, unless you explicitly whilelist these imports.
- I updated the lemma sorting heuristic to incorporate the string length of the replacement. However, I found that the previous heuristics were better, so this new heuristic is only a tie-breaker for the old heuristics.
- In RefinedDiscrTree, when entering implicit arguments, use the `instances` instead of `default` transparency. Reason: We don't want the lemma `Id.ext_iff` to show up on a goal `5 = 2`, even though it technically does apply (by secretly unifying `Nat` with `Id Nat`). The motivation for using `default` tranparency is that this matches the behaviour of `isDefEq` in reducible transparency. But matching `isDefEq` isn't necessarily a good thing. In lean's default DiscrTree, the transparency is always `reducible`. However, there is one place where `fun_prop` doesn't work if we don't have `instances` transparency. So using `instances` transparency is a good compromise.

JovanGerb commented:
The test file is pretty big and it will have to be modified when some lemmas will be added or renamed. Do you think the test should be shortened or removed for the sake of mantainability?

JovanGerb commented:
Following the developments at https://github.com/leanprover/lean4/pull/6672, I removed the "blacklisted imports" part, and instead adopted the `Lean.Name.isMetaProgramming` function. As a result, this PR now depends on the fix in #25955

fpvandoorn reviewed (approved):
It would be nice if there were tests of the form "if you click on this subexpression, you get these suggestions", but I guess we don't have that framework.

I noticed some bugs when testing.
Code:
```lean
example (n m : ℕ) : n + m = 0 := by 
  rw??
```
Click on `n + m` and select the fist suggestion (unfold). This gives an inaccessible name:
```lean
example (n m : ℕ) : n + m = 0 := by 
  rw [show n + m = n.add m from rfl✝]
```

Code:
```lean
example (n m : ℕ) : n + m + 0 = m + n := by 
  rw??
```
Click on `n + m` and select the first non-unfold suggestion. This gives
```lean
example (n m : ℕ) : n + m + 0 = m + n := by 
  nth_rw 2 [Nat.add_comm n m]
```
This fails, but either of these two would have worked:
```lean
example (n m : ℕ) : n + m + 0 = m + n := by 
  nth_rw 1 [Nat.add_comm n m]
example (n m : ℕ) : n + m + 0 = m + n := by 
  nth_rw 2 [Nat.add_comm]
```
The first of these is what the tactic should have given; the counting would have to take into account the explicit arguments you're giving to the lemma.

Otherwise, LGTM

JovanGerb commented on Mathlib/Lean/Meta/RefinedDiscrTree/Lookup.lean:
```diff
@@ -112,6 +112,13 @@ Convert a `MatchResult` into a `Array`, with better matches at the start of the
 
+/--
+Convert a `MatchResult` into an `Array` of `Array`s. Each `Array` corresponds to one pattern.
+The better matching patterns are at the start of the array.
```
The inner arrays appear in the order in which they were inserted into the discrimination tree. I can add a comment about it.

fpvandoorn replied:
Yes, a brief comment would be good.

JovanGerb commented:
In your example
```
example (n m : ℕ) : n + m + 0 = m + n := by 
  rw??
```
You found a very niche bug! The code that determines the position for `nth_rw` looks for all occurrences of `n + m` in the expression. In this case it found `n + m + 0` (!!) and `n + m`. The issue is that I forgot to wrap this in a `withReducible`. So after this fix, it will give `rw [Nat.add_comm n m]` or `rw [Nat.add_comm m n]` depending on which expression was selected.

Some considerations about how to determine the position for `nth_rw`:
- Separately computing the position for each rewrite takes a bit too long (unless we can do some java script trickery to only compute this after clicking on the suggestion)
- The current approach is to compute the position only once, independent of the rewrite lemma. Then we explicitly pass the arguments to the rewrite lemma (e.g. `Nat.add_comm n m`). Disadvantage: this doesn't work for lemmas like `and_comm` which don't have explicit arguments.
- Another approach could be to determine the positions for each pattern (i.e. each block of rewrite suggestions) separately, and then we do not pass any arguments to the rewrite lemma. Although not perfect (lemmas with the same pattern can have different type class assumptions), this would be a lot more accurate.

So a question here is: do we want `rw??` to fill in the arguments to the lemma or not?

fpvandoorn commented:
Ah, interesting. I did not expect it relied on unfolding the definition of `+`.

I do not feel strongly about passing arguments or not. I have weak preference for "yes", but if the implementation is more reliably correct with "no", then that seems fine.

(Btw, I also do not care strongly about suggesting `nth_rw`, and I'm happy if it rewrites more occurrences as well. This *does* require supplying arguments, since without supplying arguments, a different position could be rewritten, and the selected positions could be ignored.)

One other option, but I don't know how feasible this is: also give implicit arguments explicit, using the named variables (e.g. `rw [and_comm (a := True) (b := False)]`). Only do this for implicit arguments that are not type-class arguments, and do not occur in the type of a later argument.

JovanGerb commented:
For filtering the declarations, I rely on `LazyDiscrTree.blacklistInsertion`. It filters out most automatically generated things for inductives, but for some reason `Bla.mk.sizeOf_spec` and `Bla.mk.injEq` are the exceptions. So I should probably just write a more complete version of `LazyDiscrTree.blacklistInsertion`.

JovanGerb commented:
I've reduced the import and the size of the test, so that it'll be manageable to maintain. It might also be useful to have a performance test for `rw??` (to test both the initialization of RefinedDiscrTree, and the actual `rw` search), but I'm not sure whether and in what form this should be added.

fpvandoorn commented on MathlibTest/LibraryRewrite.lean:
```diff
@@ -267,242 +116,99 @@ Pattern x / y
-Pattern ∑ x ∈ s, (f x + g x)
-· ∑ x ∈ Finset.range n, x + ∑ x ∈ Finset.range n, 1
-  Finset.sum_add_distrib
+def AtZero (f : Int → Int) : Int := f 0
```
```suggestion
def atZero (f : Int → Int) : Int := f 0
```
Not at all important in tests, but it's hard to unsee.

JovanGerb commented:
bors r+
